## Import the Raw Data

Having explored the initial source file provided by Jack Chang in Kaggle entitled 'telco_churn.csv', it's time to start cleaning or pre-processing the dataset. This is probably one of the more time consuming elements creating a pipeline but is an essential phaze before performing EDA.

In [1]:
# Import all necessary libraries and modules
import numpy as np
import pandas as pd

Read the CSV source file into a dataframe and assign it to a variable called telco_churn.

In [2]:
telco_churn = pd.read_csv("C:/Users/lynst/Documents/Datasets/Kaggle/Jack Chang/telco_churn.csv")
telco_churn.head()

CustomerID  Count        Country       State         City  Zip Code  \
0  3668-QPYBK      1  United States  California  Los Angeles     90003   
1  9237-HQITU      1  United States  California  Los Angeles     90005   
2  9305-CDSKC      1  United States  California  Los Angeles     90006   
3  7892-POOKP      1  United States  California  Los Angeles     90010   
4  0280-XJGEX      1  United States  California  Los Angeles     90015   

                 Lat Long   Latitude   Longitude  Gender  ...        Contract  \
0  33.964131, -118.272783  33.964131 -118.272783    Male  ...  Month-to-month   
1   34.059281, -118.30742  34.059281 -118.307420  Female  ...  Month-to-month   
2  34.048013, -118.293953  34.048013 -118.293953  Female  ...  Month-to-month   
3  34.062125, -118.315709  34.062125 -118.315709  Female  ...  Month-to-month   
4  34.039224, -118.266293  34.039224 -118.266293    Male  ...  Month-to-month   

  Paperless Billing             Payment Method  Monthly Charges Total Charges  \
0               Yes               Mailed check            53.85        108.15   
1               Yes           Electronic check            70.70        151.65   
2               Yes           Electronic check            99.65         820.5   
3               Yes           Electronic check           104.80       3046.05   
4               Yes  Bank transfer (automatic)           103.70        5036.3   

  Churn Label Churn Value Churn Score  CLTV                   Churn Reason  
0         Yes           1          86  3239   Competitor made better offer  
1         Yes           1          67  2701                          Moved  
2         Yes           1          86  5372                          Moved  
3         Yes           1          84  5003                          Moved  
4         Yes           1          89  5340  Competitor had better devices  

[5 rows x 33 columns]

## Data Pre-processing and Cleaning

### Dimensionality Reduction

First I need to remove the columns I don't want using the drop( ) method. Be careful here as any attempt to assign this operation to the existing 'telco_churn' variable name will result in an error.

In [3]:
telco_churn.drop(['CustomerID','Count','Country','Latitude','Longitude','State','Churn Label'], axis=1, inplace=True)

telco_churn.head()

City  Zip Code                Lat Long  Gender Senior Citizen  \
0  Los Angeles     90003  33.964131, -118.272783    Male             No   
1  Los Angeles     90005   34.059281, -118.30742  Female             No   
2  Los Angeles     90006  34.048013, -118.293953  Female             No   
3  Los Angeles     90010  34.062125, -118.315709  Female             No   
4  Los Angeles     90015  34.039224, -118.266293    Male             No   

  Partner Dependents  Tenure Months Phone Service Multiple Lines  ...  \
0      No         No              2           Yes             No  ...   
1      No        Yes              2           Yes             No  ...   
2      No        Yes              8           Yes            Yes  ...   
3     Yes        Yes             28           Yes            Yes  ...   
4      No        Yes             49           Yes            Yes  ...   

  Streaming Movies        Contract Paperless Billing  \
0               No  Month-to-month               Yes   
1               No  Month-to-month               Yes   
2              Yes  Month-to-month               Yes   
3              Yes  Month-to-month               Yes   
4              Yes  Month-to-month               Yes   

              Payment Method Monthly Charges Total Charges Churn Value  \
0               Mailed check           53.85        108.15           1   
1           Electronic check           70.70        151.65           1   
2           Electronic check           99.65         820.5           1   
3           Electronic check          104.80       3046.05           1   
4  Bank transfer (automatic)          103.70        5036.3           1   

  Churn Score  CLTV                   Churn Reason  
0          86  3239   Competitor made better offer  
1          67  2701                          Moved  
2          86  5372                          Moved  
3          84  5003                          Moved  
4          89  5340  Competitor had better devices  

[5 rows x 26 columns]

In [4]:
telco_churn.shape

(7043, 26)

So there are a total of 7043 row entries or instances and 26 columns or features.

To list the names of the columns and their data type:

In [5]:
telco_churn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   City               7043 non-null   object 
 1   Zip Code           7043 non-null   int64  
 2   Lat Long           7043 non-null   object 
 3   Gender             7043 non-null   object 
 4   Senior Citizen     7043 non-null   object 
 5   Partner            7043 non-null   object 
 6   Dependents         7043 non-null   object 
 7   Tenure Months      7043 non-null   int64  
 8   Phone Service      7043 non-null   object 
 9   Multiple Lines     7043 non-null   object 
 10  Internet Service   7043 non-null   object 
 11  Online Security    7043 non-null   object 
 12  Online Backup      7043 non-null   object 
 13  Device Protection  7043 non-null   object 
 14  Tech Support       7043 non-null   object 
 15  Streaming TV       7043 non-null   object 
 16  Streaming Movies   7043 

To remove any white space in the column names using the replace method:

In [6]:
telco_churn.columns = telco_churn.columns.str.replace(' ','_')
telco_churn.columns

Index(['City', 'Zip_Code', 'Lat_Long', 'Gender', 'Senior_Citizen', 'Partner',
       'Dependents', 'Tenure_Months', 'Phone_Service', 'Multiple_Lines',
       'Internet_Service', 'Online_Security', 'Online_Backup',
       'Device_Protection', 'Tech_Support', 'Streaming_TV', 'Streaming_Movies',
       'Contract', 'Paperless_Billing', 'Payment_Method', 'Monthly_Charges',
       'Total_Charges', 'Churn_Value', 'Churn_Score', 'CLTV', 'Churn_Reason'],
      dtype='object')

Although the purpose of having unique identifiers to label each entry or customer id becomes useful when manipulating data in SQL, it will not provide any insight or potential relationships if included in a machine learning model so it was more prudent to drop the 'CustomerID' column. (In structured relational databases a unique identifier column of values becomes useful for establishing relationships in a Schema table object).

### Remove White Space
This can be achieved using the 'replace( )' method on a DataFrame.

Counting the number of unique values in the 'City' column as first.

In [7]:
telco_churn['City'].nunique()

1129

Later on a Decision Tree will be created using GraphViz but in order to draw a tree properly it is not preferable to have any whitespace in the column values for 'City', so these should be replaced with an underscore character. Looking at the first five values for 'City' using slicing.

In [8]:
telco_churn['City'].unique()[0:5]

array(['Los Angeles', 'Beverly Hills', 'Huntington Park', 'Lynwood',
       'Marina Del Rey'], dtype=object)

Now check the first 5 rows of the 'City' column using the 'head( )' method having replaced whitespaces with underscore characters.

In [9]:
telco_churn['City'].replace(' ', '_', regex=True, inplace=True)
telco_churn['City'].head()

0    Los_Angeles
1    Los_Angeles
2    Los_Angeles
3    Los_Angeles
4    Los_Angeles
Name: City, dtype: object

### Converting Data Types

Next I would like to convert any columns with string object datatypes into numeric values, all except for the 'City' column.

Gender needs converting to '1' for Male and '0' for Female for simplicity. 

In [10]:
gender_dict = {'Male': 1, 'Female': 0}

Now map the gender dictionary to the 'Gender' column:

In [11]:
telco_churn['Gender'] = telco_churn['Gender'].map(gender_dict)
telco_churn['Gender'].head()

0    1
1    0
2    0
3    0
4    1
Name: Gender, dtype: int64

'Senior Citizen' values can be converted to '1' for yes and '0' for no.

In [12]:
senior_dict = {'Yes': 1, 'No': 0}
telco_churn['Senior_Citizen'] = telco_churn['Senior_Citizen'].map(senior_dict)

Same for 'Partner':

In [13]:
partner_dict = {'Yes': 1, 'No': 0}
telco_churn['Partner'] = telco_churn['Partner'].map(partner_dict)

And 'Dependents':

In [14]:
dependents_dict = {'Yes': 1, 'No': 0}
telco_churn['Dependents'] = telco_churn['Dependents'].map(dependents_dict)

'Tenure Months' are already numeric integers so this is fine. 'Phone Service' can be converted:

In [15]:
phone_dict = {'Yes': 1, 'No': 0}
telco_churn['Phone_Service'] = telco_churn['Phone_Service'].map(phone_dict)

And 'Multiple Lines':

In [16]:
multi_dict = {'Yes': 1, 'No': 0}
telco_churn['Multiple_Lines'] = telco_churn['Multiple_Lines'].map(multi_dict)

'Internet Service' has categorical values so categorical or 'one-hot' encoding will be used here to assign different numeric values for each option. I will assign the 'DSL' value as 1, 'Fibre optic' as 2, 

In [17]:
from numpy import array
from numpy import argmax
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

data_internet = telco_churn['Internet_Service']
values_internet = array(data_internet)
print(values_internet)

['DSL' 'Fiber optic' 'Fiber optic' ... 'Fiber optic' 'DSL' 'Fiber optic']


In [18]:
# integer encode
label_encoder = LabelEncoder()
integer_encoded_internet = label_encoder.fit_transform(values_internet)
print(integer_encoded_internet)

[0 1 1 ... 1 0 1]


In [19]:
# encode the values from internet service to numbers
onehot_encoder = OneHotEncoder(sparse=False)
integer_encoded_internet = integer_encoded_internet.reshape(len(integer_encoded_internet), 1)
onehot_encoded_internet_service = onehot_encoder.fit_transform(integer_encoded_internet)
print(onehot_encoded_internet_service)

C:\Users\lynst\anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


[[1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 ...
 [0. 1. 0.]
 [1. 0. 0.]
 [0. 1. 0.]]


Test the first value.

In [20]:
# invert first example
inverted = label_encoder.inverse_transform([argmax(onehot_encoded_internet_service[0, :])])
print(inverted)

['DSL']


Repeating for the 'Contract' column which is also a categorical feature:

In [21]:
data_contract = telco_churn['Contract']
values_contract = array(data_contract)
print(values_contract)

['Month-to-month' 'Month-to-month' 'Month-to-month' ... 'One year'
 'Month-to-month' 'Two year']


In [22]:
# integer encode
label_encoder = LabelEncoder()
integer_encoded_contract = label_encoder.fit_transform(values_contract)
print(integer_encoded_contract)

[0 0 0 ... 1 0 2]


In [23]:
# encode the values from internet service to numbers
onehot_encoder = OneHotEncoder(sparse=False)
integer_encoded_contract = integer_encoded_contract.reshape(len(integer_encoded_contract), 1)
onehot_encoded_contract = onehot_encoder.fit_transform(integer_encoded_contract)
print(onehot_encoded_contract)

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 ...
 [0. 1. 0.]
 [1. 0. 0.]
 [0. 0. 1.]]


C:\Users\lynst\anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Test the first value again.

In [24]:
# invert first example
inverted = label_encoder.inverse_transform([argmax(onehot_encoded_internet_service[0, :])])
print(inverted)

['Month-to-month']


In [25]:
security_dict = {'Yes': 1, 'No': 0}
telco_churn['Online_Security'] = telco_churn['Online_Security'].map(security_dict)

In [26]:
backup_dict = {'Yes':1, 'No':0}
telco_churn['Online_Backup'] = telco_churn['Online_Backup'].map(backup_dict)

In [27]:
device_protection_dict = {'Yes': 1, 'No': 0}
telco_churn['Device_Protection'] = telco_churn['Device_Protection'].map(device_protection_dict)

In [28]:
tech_support_dict = {'Yes': 1, 'No': 0}
telco_churn['Tech_Support'] = telco_churn['Tech_Support'].map(tech_support_dict)

In [29]:
stream_tv_dict = {'Yes': 1, 'No': 0}
telco_churn['Streaming_TV'] = telco_churn['Streaming_TV'].map(stream_tv_dict)

In [30]:
stream_movies_dict = {'Yes': 1, 'No': 0}
telco_churn['Streaming_Movies'] = telco_churn['Streaming_Movies'].map(stream_movies_dict)

In [31]:
data = telco_churn['Contract']
values = array(data)
print(values)

['Month-to-month' 'Month-to-month' 'Month-to-month' ... 'One year'
 'Month-to-month' 'Two year']


In [32]:
paperless_bill_dict = {'Yes': 1, 'No': 0}
telco_churn['Paperless_Billing'] = telco_churn['Paperless_Billing'].map(paperless_bill_dict)

'Payment_Method' also has categorical features which can be converted using one-hot encoding.

In [33]:
data = telco_churn['Payment_Method']
values = array(data)
print(values)

['Mailed check' 'Electronic check' 'Electronic check' ...
 'Credit card (automatic)' 'Electronic check' 'Bank transfer (automatic)']


In [34]:
# integer encode
label_encoder = LabelEncoder()
integer_encoded_payment = label_encoder.fit_transform(values)
print(integer_encoded_payment)

[3 2 2 ... 1 2 0]


In [35]:
# binary encode
onehot_encoder = OneHotEncoder(sparse=False)
integer_encoded_payment = integer_encoded_payment.reshape(len(integer_encoded_payment), 1)
onehot_encoded_payment = onehot_encoder.fit_transform(integer_encoded_payment)
print(onehot_encoded_payment)

[[0. 0. 0. 1.]
 [0. 0. 1. 0.]
 [0. 0. 1. 0.]
 ...
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [1. 0. 0. 0.]]


C:\Users\lynst\anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


'Monthly_Charges' are already of float64 datatype, but 'Total_Charges' aren't so they need to be converted from string object to float64.

In [36]:
telco_churn['Total_Charges'] = pd.to_numeric(telco_churn['Total_Charges'], errors='coerce')
telco_churn['Total_Charges'].head()

0     108.15
1     151.65
2     820.50
3    3046.05
4    5036.30
Name: Total_Charges, dtype: float64

The 'Churn_Value' will be designated as the target vector for the purpose of my model and it currently has a simple integer datatype labeled as '1' for churn and '0' for no churn. I can already use these labels.

'Churn_Score' remains as an integer. It should be highly correlated with 'Churn_Value' so I will check for this next.

Finally, 'CLTV' which is the Customer Lifetime Value is an indicator of the customers importance during this period. This value should be compared to 'Churn_Value' or 'Churn_Score' for any association.

In addition to converting all the binary decision outcomes for certain columns above, it may be more prudent to create a for loop to iterate through these columns in a subset to speed up the algorithm.

In [37]:
# only include all columns with binary outcomes to be converted to integer values
# telco_subset = telco_churn['Gender','Senior_Citizen','Partner','Dependents', 'Phone_Service','Multiple_Lines','Online_Security','Online_Backup','Device_Protection','Tech_Support','Streaming_TV','Streaming_Movies','Paperless_Billing']
# i=0
# define a new function to iterate through the data values and change them
# for i in telco_subset:
    # dict = {'Yes': 1, 'No': 0}
    # telco_subset = telco_subset.map(dict)
    # i+=1

### Identify Missing Values
Looking for missing values and removing them is the next phase, although it may be better to replace them with 0, or impute average values. First I want to calculate the number of entries.

In [38]:
telco_churn.index

RangeIndex(start=0, stop=7043, step=1)

There are a total of 7043 rows in the dataset. Having removed the 'CustomerID' column I want to view all the columns in the dataframe once more.

In [39]:
telco_churn.columns

Index(['City', 'Zip_Code', 'Lat_Long', 'Gender', 'Senior_Citizen', 'Partner',
       'Dependents', 'Tenure_Months', 'Phone_Service', 'Multiple_Lines',
       'Internet_Service', 'Online_Security', 'Online_Backup',
       'Device_Protection', 'Tech_Support', 'Streaming_TV', 'Streaming_Movies',
       'Contract', 'Paperless_Billing', 'Payment_Method', 'Monthly_Charges',
       'Total_Charges', 'Churn_Value', 'Churn_Score', 'CLTV', 'Churn_Reason'],
      dtype='object')

In [40]:
telco_missing = pd.isnull(telco_churn).sum()
print(telco_missing)

City                    0
Zip_Code                0
Lat_Long                0
Gender                  0
Senior_Citizen          0
Partner                 0
Dependents              0
Tenure_Months           0
Phone_Service           0
Multiple_Lines        682
Internet_Service        0
Online_Security      1526
Online_Backup        1526
Device_Protection    1526
Tech_Support         1526
Streaming_TV         1526
Streaming_Movies     1526
Contract                0
Paperless_Billing       0
Payment_Method          0
Monthly_Charges         0
Total_Charges          11
Churn_Value             0
Churn_Score             0
CLTV                    0
Churn_Reason         5174
dtype: int64


Within the dataframe I need to count the location of the rows where the 'Total_Charges' column which have no entry is identical to 'True'. This counts the number of blank spaces in the particular column.

In [41]:
len(telco_churn.loc[telco_churn['Total_Charges'] == ' '])

0

So this means that none of the entries have any blank spaces. Try printing these entries out.

In [42]:
# print these rows
telco_churn.loc[telco_churn['Total_Charges'] == ' ']

Empty DataFrame
Columns: [City, Zip_Code, Lat_Long, Gender, Senior_Citizen, Partner, Dependents, Tenure_Months, Phone_Service, Multiple_Lines, Internet_Service, Online_Security, Online_Backup, Device_Protection, Tech_Support, Streaming_TV, Streaming_Movies, Contract, Paperless_Billing, Payment_Method, Monthly_Charges, Total_Charges, Churn_Value, Churn_Score, CLTV, Churn_Reason]
Index: []

[0 rows x 26 columns]

Perhaps they contain another value such as NaN, or 0 already. 

Using the isnull( ) method I may be able to find the sum of these entries and print them out.

In [43]:
telco_churn['Total_Charges'].isnull().sum()

11

In [44]:
telco_churn[telco_churn['Total_Charges'].isnull()]

City  Zip_Code                Lat_Long  Gender  \
2234  San_Bernardino     92408  34.084909, -117.258107       0   
2438    Independence     93526  36.869584, -118.189241       1   
2568       San_Mateo     94401  37.590421, -122.306467       0   
2667       Cupertino     95014  37.306612, -122.080621       1   
2856        Redcrest     95569  40.363446, -123.835041       0   
4331     Los_Angeles     90029  34.089953, -118.294824       1   
4687        Sun_City     92585  33.739412, -117.173334       1   
5104      Ben_Lomond     95005  37.078873, -122.090386       0   
5719        La_Verne     91750  34.144703, -117.770299       1   
6772            Bell     90201  33.970343, -118.171368       0   
6840      Wilmington     90744  33.782068, -118.262263       1   

      Senior_Citizen  Partner  Dependents  Tenure_Months  Phone_Service  \
2234               0        1           0              0              0   
2438               0        0           0              0              1   
2568               0        1           0              0              1   
2667               0        1           1              0              1   
2856               0        1           0              0              0   
4331               0        1           1              0              1   
4687               0        1           1              0              1   
5104               0        1           1              0              1   
5719               0        1           1              0              1   
6772               0        1           1              0              1   
6840               0        0           1              0              1   

      Multiple_Lines  ... Streaming_Movies  Contract  Paperless_Billing  \
2234             NaN  ...              0.0  Two year                  1   
2438             0.0  ...              NaN  Two year                  0   
2568             0.0  ...              1.0  Two year                  0   
2667             1.0  ...              NaN  Two year                  0   
2856             NaN  ...              0.0  Two year                  0   
4331             0.0  ...              NaN  Two year                  0   
4687             1.0  ...              NaN  Two year                  0   
5104             0.0  ...              NaN  Two year                  0   
5719             0.0  ...              NaN  One year                  1   
6772             1.0  ...              0.0  Two year                  0   
6840             1.0  ...              0.0  Two year                  1   

                 Payment_Method  Monthly_Charges  Total_Charges  Churn_Value  \
2234  Bank transfer (automatic)            52.55            NaN            0   
2438               Mailed check            20.25            NaN            0   
2568               Mailed check            80.85            NaN            0   
2667               Mailed check            25.75            NaN            0   
2856    Credit card (automatic)            56.05            NaN            0   
4331               Mailed check            19.85            NaN            0   
4687               Mailed check            25.35            NaN            0   
5104               Mailed check            20.00            NaN            0   
5719               Mailed check            19.70            NaN            0   
6772               Mailed check            73.35            NaN            0   
6840  Bank transfer (automatic)            61.90            NaN            0   

     Churn_Score  CLTV Churn_Reason  
2234          36  2578          NaN  
2438          68  5504          NaN  
2568          45  2048          NaN  
2667          48  4950          NaN  
2856          30  4740          NaN  
4331          53  2019          NaN  
4687          49  2299          NaN  
5104          27  3763          NaN  
5719          69  4890          NaN  
6772          44  2342          NaN  
6840          65  5188          NaN  

[11

I can see all eleven Null values and decide whether to remove these rows from the dataframe completely, or impute some kind of average value or a 0. In this case they've been assigned values of 'NaN', or 'Not a Number'.

I've decided to set these missing values to 0 for now. I can always remove the values later and try running the model again to see if there is any difference to the scoring metric.

In [45]:
telco_churn.loc[(telco_churn['Total_Charges'] == 'NaN'), 'Total_Charges'] = 0

Let me see if this has worked and the 'Total_Charges' column with NaN values have been converted to 0.

In [46]:
telco_churn[telco_churn['Total_Charges'].isnull()]

City  Zip_Code                Lat_Long  Gender  \
2234  San_Bernardino     92408  34.084909, -117.258107       0   
2438    Independence     93526  36.869584, -118.189241       1   
2568       San_Mateo     94401  37.590421, -122.306467       0   
2667       Cupertino     95014  37.306612, -122.080621       1   
2856        Redcrest     95569  40.363446, -123.835041       0   
4331     Los_Angeles     90029  34.089953, -118.294824       1   
4687        Sun_City     92585  33.739412, -117.173334       1   
5104      Ben_Lomond     95005  37.078873, -122.090386       0   
5719        La_Verne     91750  34.144703, -117.770299       1   
6772            Bell     90201  33.970343, -118.171368       0   
6840      Wilmington     90744  33.782068, -118.262263       1   

      Senior_Citizen  Partner  Dependents  Tenure_Months  Phone_Service  \
2234               0        1           0              0              0   
2438               0        0           0              0              1   
2568               0        1           0              0              1   
2667               0        1           1              0              1   
2856               0        1           0              0              0   
4331               0        1           1              0              1   
4687               0        1           1              0              1   
5104               0        1           1              0              1   
5719               0        1           1              0              1   
6772               0        1           1              0              1   
6840               0        0           1              0              1   

      Multiple_Lines  ... Streaming_Movies  Contract  Paperless_Billing  \
2234             NaN  ...              0.0  Two year                  1   
2438             0.0  ...              NaN  Two year                  0   
2568             0.0  ...              1.0  Two year                  0   
2667             1.0  ...              NaN  Two year                  0   
2856             NaN  ...              0.0  Two year                  0   
4331             0.0  ...              NaN  Two year                  0   
4687             1.0  ...              NaN  Two year                  0   
5104             0.0  ...              NaN  Two year                  0   
5719             0.0  ...              NaN  One year                  1   
6772             1.0  ...              0.0  Two year                  0   
6840             1.0  ...              0.0  Two year                  1   

                 Payment_Method  Monthly_Charges  Total_Charges  Churn_Value  \
2234  Bank transfer (automatic)            52.55            NaN            0   
2438               Mailed check            20.25            NaN            0   
2568               Mailed check            80.85            NaN            0   
2667               Mailed check            25.75            NaN            0   
2856    Credit card (automatic)            56.05            NaN            0   
4331               Mailed check            19.85            NaN            0   
4687               Mailed check            25.35            NaN            0   
5104               Mailed check            20.00            NaN            0   
5719               Mailed check            19.70            NaN            0   
6772               Mailed check            73.35            NaN            0   
6840  Bank transfer (automatic)            61.90            NaN            0   

     Churn_Score  CLTV Churn_Reason  
2234          36  2578          NaN  
2438          68  5504          NaN  
2568          45  2048          NaN  
2667          48  4950          NaN  
2856          30  4740          NaN  
4331          53  2019          NaN  
4687          49  2299          NaN  
5104          27  3763          NaN  
5719          69  4890          NaN  
6772          44  2342          NaN  
6840          65  5188          NaN  

[11

In [47]:
telco_churn['Total_Charges'].unique()

array([ 108.15,  151.65,  820.5 , ..., 7362.9 ,  346.45, 6844.5 ])

In [48]:
telco_churn.columns

Index(['City', 'Zip_Code', 'Lat_Long', 'Gender', 'Senior_Citizen', 'Partner',
       'Dependents', 'Tenure_Months', 'Phone_Service', 'Multiple_Lines',
       'Internet_Service', 'Online_Security', 'Online_Backup',
       'Device_Protection', 'Tech_Support', 'Streaming_TV', 'Streaming_Movies',
       'Contract', 'Paperless_Billing', 'Payment_Method', 'Monthly_Charges',
       'Total_Charges', 'Churn_Value', 'Churn_Score', 'CLTV', 'Churn_Reason'],
      dtype='object')

In [49]:
for column in telco_churn.columns.values.tolist():
    print(column)
    print (telco_churn[column].value_counts())
    print("")

City
Los_Angeles       305
San_Diego         150
San_Jose          112
Sacramento        108
San_Francisco     104
                 ... 
Healdsburg          4
Jenner              4
Philo               4
Point_Arena         4
Olympic_Valley      4
Name: City, Length: 1129, dtype: int64

Zip_Code
90003    5
91436    5
91916    5
91913    5
91911    5
        ..
95303    4
95251    4
95250    4
95246    4
96146    4
Name: Zip_Code, Length: 1652, dtype: int64

Lat_Long
33.964131, -118.272783    5
34.152875, -118.486056    5
32.912664, -116.635387    5
32.64164, -116.985026     5
32.607964, -117.059459    5
                         ..
37.4695, -120.672724      4
38.055562, -120.456298    4
38.244806, -120.417301    4
38.264262, -120.515133    4
39.191797, -120.212401    4
Name: Lat_Long, Length: 1652, dtype: int64

Gender
1    3555
0    3488
Name: Gender, dtype: int64

Senior_Citizen
0    5901
1    1142
Name: Senior_Citizen, dtype: int64

Partner
0    3641
1    3402
Name: Partner, dtype: in

Print the cleaned dataset which has undergone the first part of the Transformation phase in the ETL pipeline.

In [50]:
print(telco_churn)

              City  Zip_Code                Lat_Long  Gender  Senior_Citizen  \
0      Los_Angeles     90003  33.964131, -118.272783       1               0   
1      Los_Angeles     90005   34.059281, -118.30742       0               0   
2      Los_Angeles     90006  34.048013, -118.293953       0               0   
3      Los_Angeles     90010  34.062125, -118.315709       0               0   
4      Los_Angeles     90015  34.039224, -118.266293       1               0   
...            ...       ...                     ...     ...             ...   
7038       Landers     92285  34.341737, -116.539416       0               0   
7039      Adelanto     92301  34.667815, -117.536183       1               0   
7040         Amboy     92304  34.559882, -115.637164       0               0   
7041  Angelus_Oaks     92305     34.1678, -116.86433       0               0   
7042  Apple_Valley     92308  34.424926, -117.184503       1               0   

      Partner  Dependents  Tenure_Month